AppTaskQueue/appscale/taskqueue/appscale_taskqueue.py

```diff
@@ -202,65 +183,129 @@ def remote_request(self, app_info, http_request_data):
+      else:
+        recent_stats = service_stats.get_recent()
+    except ValueError:
+      self.set_status(400, "cursor and last_milliseconds "
+                           "arguments should be integers")
+      return
+
+    with (yield stats_lock.acquire()):
+      cumulative_counters = service_stats.get_cumulative_counters()
+
+    tq_stats = {
+      "current_requests": service_stats.current_requests,
+      "cumulative_counters": cumulative_counters,
+      "recent_stats": recent_stats
+    }
+    self.write(json.dumps(tq_stats))
+    self.finish()
 
 
-def graceful_shutdown(*_):
-  """ Stop accepting new requests and exit on the next I/O loop iteration.
+def prepare_taskqueue_application(task_queue):
+  handlers = [
+    # Provides compatibility with the v1beta2 REST API.
+    (RESTQueue.PATH, RESTQueue, {'queue_handler': task_queue}),
+    (RESTTasks.PATH, RESTTasks, {'queue_handler': task_queue}),
+    (RESTLease.PATH, RESTLease, {'queue_handler': task_queue}),
+    (RESTTask.PATH, RESTTask, {'queue_handler': task_queue}),
+    # Responds with service statistic
+    ("/service-stats", StatsHandler),
+    # Takes protocol buffers from the AppServers.
+    (r"/.*", ProtobufferHandler, {'queue_handler': task_queue})
+  ]
+
+  return Application(handlers)
 
-  This is safe as long as the server is synchronous. It will stop in the middle
-  of requests as soon as the server has asynchronous handlers.
+
+def prepare_graceful_shutdown(zk_client, tornado_server):
+  """ Defines function which should handle termination signal.
+  
+  Args:
+    zk_client: an instance of zookeeper client.
+    tornado_server: an instance of tornado server.
+  Returns:
+    a callable which takes care about graceful shutdown.
   """
-  logger.info('Stopping server')
-  zk_client.stop()
-  server.stop()
-  io_loop = IOLoop.instance()
-  io_loop.add_callback_from_signal(io_loop.stop)
+
+  def graceful_shutdown(*_):
+    """ Stop accepting new requests and exit when all requests are finished
+    or timeout is exceeded.
+
+    This is safe as long as the server is synchronous. It will stop in the middle
+    of requests as soon as the server has asynchronous handlers.
+    """
+    signal_time = time.time()
+    logger.info('Stopping server')
+    tornado_server.stop()
+    io_loop = ioloop.IOLoop.current()
+
+    def stop_on_signal():
+      current_requests = service_stats.current_requests
+      if current_requests and time.time() - signal_time < SHUTTING_DOWN_TIMEOUT:
+        logger.warning("Can't stop Taskqueue server now as {reqs} requests are "
+                       "still in progress".format(reqs=current_requests))
+      else:
+        if current_requests:
+          logger.error("Shutting down server despite {reqs} requests "
+                       "in progress".format(reqs=current_requests))
+        io_loop.stop()     # As all requests are
+        zk_client.stop()   # The latest thing to stop is ZK client
```

Comment:
Just to be picky, "last" is probably more fitting than "latest" because "latest" implies that it's the last until now, but more things might be stopped in the future. In other words, every line is always the latest thing to happen at the moment of execution.

Also, if order is important enough here to merit a comment, it's usually best to explain the reason.